keps/sig-apps/sidecarcontainers.md

```diff
@@ -111,21 +112,45 @@ This will be useful in scenarios such as when your sidecar is a proxy so that it
 
 To solve the problem of Jobs that don't complete: When RestartPolicy!=Always if all normal containers have reached a terminal state (Succeeded for restartPolicy=OnFailure, or Succeeded/Failed for restartPolicy=Never), then all sidecar containers will be sent a SIGTERM.
 
+Sidecars are just normal containers in almost all respects, they have all the same attributes, they are included in pod state, obey pod restart policy etc. The only differences are lifecycle related.
+
 ### Implementation Details/Notes/Constraints
 
-As this is a fairly large change I think it make sense to break this proposal down and phase in more functionality as we go, potential roadmap could look like:
+The proposal can broken down into four key pieces of implementation that all relatively separate from one another:
 
-* Add sidecar field, use it for the shutdown triggering when RestartPolicy!=Always
+* Shutdown triggering for sidecars when RestartPolicy!=Always
 * Pre-stop hooks sent to sidecars before non sidecar containers
 * Sidecars are terminated after normal containers
 * Sidecars start before normal containers
 
-
 As this is a change to the Container spec we will be using feature gating, you will be required to explicitly enable this feature on the api server as recommended [here](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions).
 
+#### Kubelet Changes:
+Broad outline of what places could be modified to implement desired behaviour:
+
+###### Shutdown triggering:
+Package `kuberuntime`
+
+Modify `kuberuntime_manager.go`, function `computePodActions`. Have a check in this function that will see if all the non-sidecars had permanently exited, if true: return all the running sidecars in `ContainersToKill`.
+
+###### Sidecars terminated last
+Package `kuberuntime`
+
+Modify `kuberuntime_container.go`, function `killContainersWithSyncResult`. Break up the looping over containers so that it goes through killing the non-sidecars before terminating the sidecars.
```

Comment:
The inner loop iterates over [kubecontainer.Container](https://sourcegraph.com/github.com/kubernetes/kubernetes@493b26105789bf88575b78cce9e8474e20ae3f48/-/blob/pkg/kubelet/container/runtime.go#L248) (as opposed to `v1.Container`), which doesn't seem to have a way to tell whether the container is a sidecar. You might be able to cross-reference from the `v1.Pod` that also gets passed in, but we need to be careful to validate that that argument contains what we think it does. For example, I saw some comments that seemed to imply `pod` might be `nil` sometimes; we need to know what those cases are, and whether we can handle them properly.

It might end up being necessary to add data to `kubecontainer.Container`. I'm not sure if that represents a change to CRI? I'm also new to this code.

Reply:
oh yeah I missed that. Looking around it seems to indicate that `Pod may be nil, running pod must not be.` If the pod is dead we're not as bothered about term ordering, so cross referencing may work with a `if pod != nil`.

Reply:
If we can verify that `pod` is only ever `nil` when the `Pod` object on the API server has been force-deleted, then I agree it's reasonable to say, "We don't respect sidecar termination order if you force-delete the Pod." Have you found any evidence to say exactly when `pod` could be `nil`? For example, is it possible as a transient condition during a Node restart, before it fetches Pods from the API server?

Reply:
I'm not confident from looking at the code that I could tell you all the scenarios where `pod` could be `nil`. If @feiskyer could weigh in it would be appreciated.  If there are scenarios where it can be nil and we still care about ordering then we'll need to look at a different way to enable this behaviour. Changing `kubecontainer.Container` is a possibility but it means changing the container runtime.